### PR TITLE
Update to the next version of proc-macro2

### PR DESCRIPTION
**Pull request by alexcrichton**
~~Depends on dtolnay/quote#73~~
~~Depends on alexcrichton/proc-macro2#90~~
~~Depends on a new nightly~~

### CODE REVIEW

File: src/synom.rs [hunk 1]
```diff
@@ -213,6 +213,16 @@ impl Synom for proc_macro2::TokenStream {
     }
 }
 
+impl Synom for proc_macro2::Ident {
+    fn parse(input: Cursor) -> PResult<Self> {
+        input.term().ok_or_else(|| ParseError::new("not an ident"))
```
Comment (dtolnay) [inline src/synom.rs]: I would like this to preserve the behavior of 0.13's impl Synom for Ident. Separately we will need some way to do https://docs.rs/syn/0.13/syn/struct.Ident.html#method.parse_any.

File: src/synom.rs [hunk 2]
```diff
@@ -213,6 +213,16 @@ impl Synom for proc_macro2::TokenStream {
...
+    }
+
+    fn description() -> Option<&'static str> {
+        Some("arbitrary token stream")
```
Comment (dtolnay) [inline src/synom.rs]: This should be Some("identifier").

**Comment by alexcrichton** [comment]
Ok it looks like the last failure is 1.15.1 but that's just because it doesn't understand `[patch]` and should be fixable with a new release of `quote`

File: codegen/src/main.rs [hunk 3]
```diff
@@ -106,7 +107,7 @@ fn load_file<P: AsRef<Path>>(name: P, features: &Tokens, lookup: &mut Lookup) ->
                 // We don't want to try to load the generated rust files and
                 // parse them, so we ignore them here.
                 for name in IGNORED_MODS {
-                    if item.ident == name {
+                    if item.ident.to_string() == *name {
```
Comment (dtolnay) [inline codegen/src/main.rs]: I opened https://github.com/alexcrichton/proc-macro2/pull/92 to avoid these changes.

File: src/lifetime.rs [hunk 4]
```diff
@@ -140,11 +149,13 @@ pub mod parsing {
...
     use super::*;
-    use quote::{ToTokens, Tokens};
+    use quote::{ToTokens, TokenStreamExt};
+    use proc_macro2::{TokenStream, Punct, Spacing};
 
     impl ToTokens for Lifetime {
-        fn to_tokens(&self, tokens: &mut Tokens) {
-            self.term.to_tokens(tokens);
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.append(Punct::new('\'', Spacing::Joint));
```
Comment (dtolnay) [inline src/lifetime.rs]: This token needs to be stored in the Lifetime struct so that we preserve its Span correctly.

File: src/ident.rs [hunk 5]
```diff
@@ -1,306 +0,0 @@
...
-/// if ident_str.len() > 60 {
-///     println!("Very long identifier: {}", ident_str)
-/// }
-///
-/// // Create a String from the ident.
-/// let ident_string = ident.to_string();
-/// give_away(ident_string);
-///
-/// fn give_away(s: String) { /* ... */ }
-/// ```
```
Comment (dtolnay) [inline src/ident.rs]: There is valuable explanation and examples being removed here. Compare:

- https://docs.rs/syn/0.13/syn/struct.Ident.html vs
- https://docs.rs/proc-macro2/0.4.1/proc_macro2/struct.Ident.html

Any thoughts on how to provide Syn-specific Ident documentation?

File: src/gen/fold.rs [hunk 6]
```diff
@@ -191,8 +192,6 @@ fn fold_generic_method_argument(&mut self, i: GenericMethodArgument) -> GenericM
 fn fold_generic_param(&mut self, i: GenericParam) -> GenericParam { fold_generic_param(self, i) }
 # [ cfg ( any ( feature = "full" , feature = "derive" ) ) ]
 fn fold_generics(&mut self, i: Generics) -> Generics { fold_generics(self, i) }
-
-fn fold_ident(&mut self, i: Ident) -> Ident { fold_ident(self, i) }
```
Comment (dtolnay) [inline src/gen/fold.rs]: I would like to keep fold_ident.

**Comment by dtolnay** [comment]
I can't comment on it directly but please remove the tests/rust submodule added in this PR. IIRC we used to have it but it slows down Cargo git dependencies on Syn.

File: src/gen/fold.rs (hunk 6 above)
Comment (alexcrichton) [reply src/gen/fold.rs]: Oops didn't even realize that went away

File: src/ident.rs (hunk 5 above)
Comment (alexcrichton) [reply src/ident.rs]: Hm I wonder if we could just wholesale move the documentation to `proc_macro2::Ident`?

File: src/ident.rs [hunk 7]
```diff
@@ -1,306 +0,0 @@
...
-        fn description() -> Option<&'static str> {
-            Some("identifier")
-        }
-    }
-    
-    impl Ident {
-        /// Parses any identifier
-        /// 
-        /// This is useful when parsing a DSL which allows Rust keywords as identifiers.
-        pub fn parse_any(input: Cursor) -> PResult<Self> {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/ident.rs]: Perhaps a new macro parser in synom, like `any_ident!()`?

**Comment by dtolnay** [comment]
I think we are mostly waiting for @Eijebong to wrap up work on https://github.com/servo/servo/issues/20497 so that the timing can be maximally bad.

After that, docs need updating, we need some design work around how to handle what used to be `Ident::parse_any`, and I would like to get serde_derive updated before publishing so that we understand practical implications of the changes. If you have bandwidth to help, the best thing would be to make a PR for Serde with some `[patch.crates-io]` dependencies on this branch.

File: src/ident.rs (hunk 5 above)
Comment (dtolnay) [reply src/ident.rs]: Sounds good, I moved this to proc-macro2 in https://github.com/alexcrichton/proc-macro2/commit/8b71daca2dd228053ac4df17a356f667332158fd.

File: src/ident.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (dtolnay) [reply src/ident.rs]: I implemented this as an extension trait in fe68c90ec8af31a7bbbf554405c1b7e34ae4b0ed.

**Comment by dtolnay** [comment]
Serde update: https://github.com/serde-rs/serde/pull/1273

**Pull request merged**